src/config.rs

```diff
@@ -180,12 +210,142 @@ pub fn parse_config() -> Result<Config, Error> {
+
+    let proxy_config_env: anyhow::Result<serde_yaml::Value> =
+        parse_default::<String>(PROXY_CONFIG, "".to_string())
+            .map_err(anyhow::Error::new)
+            .and_then(|env_str| serde_yaml::from_str(&env_str).map_err(anyhow::Error::new));
+
+    let yaml = match (mesh_config, proxy_config_env) {
+        (Ok(mc), Ok(pc_env)) => {
+            let mut mc = mc
+                .get("defaultConfig")
+                .unwrap_or(&serde_yaml::Value::Null)
+                .clone();
+            merge_yaml(&mut mc, pc_env);
+            Ok(mc)
+        }
+        (Ok(mc), Err(_)) => Ok(mc
+            .get("defaultConfig")
+            .unwrap_or(&serde_yaml::Value::Null)
+            .clone()),
+        (Err(_), Ok(pc_env)) => Ok(pc_env),
+        (Err(mc), Err(pc_env)) => Err(anyhow!("neither parsed: {mc}; {pc_env}")),
+    };
+
+    let pc = match yaml {
+        Ok(yaml) => serde_yaml::from_value(yaml).map_err(anyhow::Error::new)?,
+        Err(e) => {
+            warn!("{}", e);
+            default_proxy_config()
+        }
+    };
+
+    // TODO add all env with ISTIO_META_ prefix
+    // TODO remove ISTIO_META_ prefix from proxy_metadata
+
+    Ok(pc)
+}
+
+fn merge_yaml(a: &mut serde_yaml::Value, b: serde_yaml::Value) {
+    match (a, b) {
+        (a @ &mut serde_yaml::Value::Mapping(_), serde_yaml::Value::Mapping(b)) => {
+            let a = a.as_mapping_mut().unwrap();
+            for (k, v) in b {
+                if v.is_sequence() && a.contains_key(&k) && a[&k].is_sequence() {
+                    let mut _b = a.get(&k).unwrap().as_sequence().unwrap().to_owned();
+                    _b.append(&mut v.as_sequence().unwrap().to_owned());
+                    a[&k] = serde_yaml::Value::from(_b);
+                    continue;
+                }
+                if !a.contains_key(&k) {
+                    a.insert(k.to_owned(), v.to_owned());
+                } else {
+                    merge_yaml(&mut a[&k], v);
+                }
+            }
+        }
+        (a, b) => *a = b,
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::{construct_config, construct_proxy_config, default_proxy_config};
+    use crate::xds::istio::mesh::ProxyConfig;
+
+    #[test]
+    fn config_from_proxyconfig() {
+        // mesh config only
+        let mesh_config_path = "./src/test_helpers/mesh_config.yaml";
+        let pc = construct_proxy_config(mesh_config_path).unwrap();
+        let cfg = construct_config(pc).unwrap();
+        assert_eq!(cfg.readiness_addr.port(), 15888);
+        assert_eq!(cfg.admin_addr.port(), 15099);
+        // TODO remove prefix
+        assert_eq!(cfg.proxy_metadata["ISTIO_META_FOO"], "bar");
+
+        // env only
+        let pc_env = "{ \
+            \"discoveryAddress\": \"istiod-rev0.istio-system-2:15012\", \
+            \"proxyAdminPort\": 15999 \
+        }";
+        std::env::set_var("PROXY_CONFIG", pc_env);
```

Comment:
do we need to unset it after?

Reply:
I'll just change the fn to accept both inputs as params to make it more testable